### PR TITLE
feat(core,phrases): validate there must be one and only one primary sign-in method

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

Validate there must be one and only one primary sign-in method

#477 → #475 → #480 → …

## Linear Issue Reference

LOG-2059

## Testing

image

### CODE REVIEW

File: packages/core/src/utils/validate-sign-in-experience.ts [hunk 1]
```diff
@@ -19,3 +25,26 @@ export const validateTermsOfUse = (termsOfUse: Optional<TermsOfUse>) => {
...
+
+export const validateSignInMethods = (signInMethods?: SignInMethods) => {
+  if (!signInMethods) {
+    return;
+  }
+
+  validateOneAndOnlyOnePrimarySignInMethod(signInMethods);
+
+  // TODO: assert others next PR
+};
```
Comment (simeng-li) [inline packages/core/src/utils/validate-sign-in-experience.ts:50]: Could merge as one single`validateSignInMethods` ?

File: packages/core/src/utils/validate-sign-in-experience.ts [hunk 2]
```diff
@@ -19,3 +25,26 @@ export const validateTermsOfUse = (termsOfUse: Optional<TermsOfUse>) => {
...
   );
 };
+
+const validateOneAndOnlyOnePrimarySignInMethod = (signInMethods: SignInMethods) => {
+  const signInMethodStates = Object.values(signInMethods);
+  const primarySignInMethodCount = signInMethodStates.reduce(
+    (primaryCount, signInMethodState) =>
+      primaryCount + (signInMethodState === SignInMethodState.primary ? 1 : 0),
+    0
+  );
```
Comment (simeng-li) [inline packages/core/src/utils/validate-sign-in-experience.ts:35]: nit: signInMethodStates.filer(state => state === SignInMethodState.primary).length

File: packages/core/src/utils/validate-sign-in-experience.ts (hunk 1 above)
Comment (IceHe) [reply packages/core/src/utils/validate-sign-in-experience.ts:50]: there will five more validate function here and this function will be too long and not readable enough to understand quickly. we can decide whether to merge later next pr.

Comment (IceHe) [reply packages/core/src/utils/validate-sign-in-experience.ts]: Inlined. PTAL.

File: packages/core/src/utils/validate-sign-in-experience.test.ts [hunk 3]
```diff
@@ -39,3 +43,36 @@ describe('validate terms of use', () => {
     }).toMatchError(new RequestError('sign_in_experiences.empty_content_url_of_terms_of_use'));
   });
 });
+
+describe('validate sign-in methods', () => {
+  test('should not throw if without sign-in methods', () => {
+    expect(() => {
+      // eslint-disable-next-line unicorn/no-useless-undefined
+      validateSignInMethods(undefined);
```
Comment (wangsijie) [inline packages/core/src/utils/validate-sign-in-experience.test.ts:51]: wondering why put an `undefined` here?

**Review by wangsijie — APPROVED** [review]

File: packages/core/src/utils/validate-sign-in-experience.test.ts (hunk 3 above)
Comment (IceHe) [reply packages/core/src/utils/validate-sign-in-experience.test.ts:51]: - function requires caller provide parameter.
- so parameter type is `xxx | undefined` and not `?: xxx`

Comment (wangsijie) [reply packages/core/src/utils/validate-sign-in-experience.test.ts:51]: There are no required argument in this function?

Comment (IceHe) [reply packages/core/src/utils/validate-sign-in-experience.test.ts:51]: Updated, PTAL again

```ts
export const validateSignInMethods = (signInMethods: SignInMethods | undefined) => {
    …
}
```

**Review by simeng-li — APPROVED** [review]
LGTM

**Pull request merged**